beacon-chain/blockchain/service.go

```diff
@@ -97,14 +97,19 @@ func NewService(ctx context.Context, opts ...Option) (*Service, error) {
 		boundaryRoots:        [][32]byte{},
 		checkpointStateCache: cache.NewCheckpointStateCache(),
 		initSyncBlocks:       make(map[[32]byte]block.SignedBeaconBlock),
-		justifiedBalances:    make([]uint64, 0),
 		cfg:                  &config{},
 	}
 	for _, opt := range opts {
 		if err := opt(srv); err != nil {
 			return nil, err
 		}
 	}
+	if srv.justifiedBalances == nil {
+		if srv.cfg.StateGen == nil {
```

Comment:
Any reasons not having the nil check inside `newStateBalanceCache` and change the signatures to
`newStateBalanceCache(sg *stategen.State) (*stateBalanceCache, error) `

Reply:
i like that idea, the whole point of the constructor is to hint that the zero-value for the cache doesn't work, so it makes sense to make it more defensive